drivers/i2c/i2c_mchp_mss.c

```diff
@@ -73,163 +72,170 @@ LOG_MODULE_REGISTER(i2c_mchp, CONFIG_I2C_LOG_LEVEL);
+			return -EINVAL;
+		}
 
-static int mss_i2c_write(const struct device *dev, uint8_t serial_addr, uint8_t *tx_buffer,
-			 uint32_t tx_num_write)
-{
-	struct mss_i2c_data *const data = dev->data;
-	const struct mss_i2c_config *cfg = dev->config;
-	uint8_t ctrl = sys_read8(cfg->i2c_base_addr + CORE_I2C_CTRL);
+		if ((curr->flags & I2C_MSG_STOP) != 0) {
+			/* Stop condition is only allowed on last message */
+			return -EINVAL;
+		}
 
-	atomic_t target_status = data->target_status;
+		if ((curr->flags & I2C_MSG_RW_MASK) == (next->flags & I2C_MSG_RW_MASK)) {
+			if ((next->flags & I2C_MSG_RESTART) != 0) {
+				/*
+				 * Restart condition between messages of the same direction
+				 * is not supported.
+				 */
+				return -EINVAL;
+			}
+		} else if ((next->flags & I2C_MSG_RESTART) == 0) {
+			/*
+			 * Restart condition between messages of different directions
+			 * is required.
+			 */
+			return -EINVAL;
+		}
 
-	if (data->transaction == NO_TRANSACTION) {
-		data->transaction = CONTROLLER_WRITE_TRANSACTION;
+		curr = next;
 	}
 
-	data->pending_transaction = CONTROLLER_WRITE_TRANSACTION;
-	data->target_addr = serial_addr << TARGET_ADDR_SHIFT;
-	data->dir = I2C_MSG_WRITE;
-	data->controller_tx_buffer = tx_buffer;
-	data->controller_tx_size = tx_num_write;
-	data->controller_tx_idx = 0u;
-	atomic_set(&data->controller_status, MSS_I2C_IN_PROGRESS);
-
-	if (target_status == MSS_I2C_IN_PROGRESS) {
-		data->is_transaction_pending = CONTROLLER_WRITE_TRANSACTION;
-	} else {
-		sys_write8((ctrl | CTRL_STA), cfg->i2c_base_addr + CORE_I2C_CTRL);
+	/* Add RW bit to address, so that we can write it directly to the data register */
+	addr <<= 1;
+	if ((msgs->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ) {
+		addr |= 1;
 	}
 
-	if (data->bus_status == MSS_I2C_HOLD_BUS) {
-		sys_write8((ctrl & ~CTRL_SI), cfg->i2c_base_addr + CORE_I2C_CTRL);
-	}
+	(void)k_mutex_lock(&data->mtx, K_FOREVER);
 
-	return 0;
-}
+	data->ret = 0;
+	data->addr = (uint8_t)addr;
+	data->msg_curr = msgs;
+	data->msg_last = msgs + num_msgs - 1;
+	data->byte_curr = mss_i2c_set_byte_end(data, msgs);
 
+	/* Start the transfer */
+	uint8_t ctrl = sys_read8(cfg->i2c_base_addr + CORE_I2C_CTRL);
 
-static int mss_i2c_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
-			    uint16_t addr)
-{
-	for (int i = 0; i < num_msgs; i++) {
-		struct i2c_msg *current = &msgs[i];
-
-		if ((current->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ) {
-			mss_i2c_read(dev, addr, current->buf, current->len);
-			mss_wait_complete(dev);
-		} else {
-			mss_i2c_write(dev, addr, current->buf, current->len);
-			mss_wait_complete(dev);
-		}
+	ctrl &= ~(CTRL_AA | CTRL_SI | CTRL_STO);
+	ctrl |= CTRL_STA;
+	sys_write8(ctrl, cfg->i2c_base_addr + CORE_I2C_CTRL);
+
+	/*
+	 * Clear a potentially erroneous done condition caused by a spurious interrupt.  Enable
+	 * interrupts and wait for the transfer completion.
+	 */
+	(void)k_sem_take(&data->done, K_NO_WAIT);
```

Comment:
k_sem_reset is fine instead

Reply:
I updated the commit to use k_sem_reset(). It is definitely easier to understand.